### PR TITLE
Symintify numel(), infer_size, prims.elementwise_meta

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89111
* __->__ #88956

### CODE REVIEW

File: aten/src/ATen/ExpandUtils.cpp [hunk 1]
```diff
@@ -25,8 +25,8 @@ Container infer_size_impl(IntArrayRef a, IntArrayRef b) {
     ptrdiff_t offset = ndim - 1 - i;
     ptrdiff_t dimA = dimsA - 1 - offset;
     ptrdiff_t dimB = dimsB - 1 - offset;
-    int64_t sizeA = (dimA >= 0) ? a[dimA] : 1;
-    int64_t sizeB = (dimB >= 0) ? b[dimB] : 1;
```
Comment (Skylion007) [inline aten/src/ATen/ExpandUtils.cpp:29]: Since these can be symInts now, you should call std::move when they are stored in the output below.

**Comment by ezyang** [comment]
Would appreciate a test for the new number promotion logic

File: aten/src/ATen/ExpandUtils.cpp (hunk 1 above)
Comment (Skylion007) [reply aten/src/ATen/ExpandUtils.cpp:29]: Specifically:
```
      expandedSizes[i] = sizeA == 1 ? std::move(sizeB) : std::move(sizeA);
```

**Review by ezyang — APPROVED** [review]

File: test/test_dynamic_shapes.py [hunk 2]
```diff
@@ -291,6 +291,17 @@ def test_size_expressions(self):
         self.assertTrue(str(expand_x.shape[1]), str(x.shape[0]))
         self.assertTrue(str(expand_x.shape[1]), str(result.shape[0]))
 
+    @skipIfNoSympy
```
Comment (SherlockNoMad) [inline test/test_dynamic_shapes.py:294]: Test added for numel

File: torch/csrc/autograd/python_torch_functions_manual.cpp [hunk 3]
```diff
@@ -692,7 +692,7 @@ static PyObject* THPVariable_numel(
   }
 
   if (r.idx == 0) {
-    return wrap(r.tensor(0).numel());
+    return py::cast(r.tensor(0).sym_numel()).release().ptr();
```
Comment (SherlockNoMad) [inline torch/csrc/autograd/python_torch_functions_manual.cpp:695]: Somehow wrap() was converting c10::SymInt into Tensor. 

We explicitly convert from c10::SymInt to python torch.SymInt.

File: test/test_dynamic_shapes.py (hunk 2 above)
Comment (SherlockNoMad) [reply test/test_dynamic_shapes.py:294]: @ezyang, I suppose this is the desirable behavior: 
numel(SymbolicTensor) returns a SymInt. 
num(RealTensor) returns an int.

File: torch/csrc/autograd/python_torch_functions_manual.cpp (hunk 3 above)
Comment (SherlockNoMad) [reply torch/csrc/autograd/python_torch_functions_manual.cpp:695]: @ezyang, fix copied from THPVariable_numel() in python_variable_methods.cpp

File: test/test_dynamic_shapes.py (hunk 2 above)
Comment (ezyang) [reply test/test_dynamic_shapes.py:294]: Yup that sounds right

**Review by ezyang — APPROVED** [review]
I wonder if wraps behavior is ever desired

**Pull request closed without merging**